network/netcore/src/transport/tcp.rs

```diff
@@ -163,6 +185,43 @@ async fn resolve_and_connect(addr: NetworkAddress) -> io::Result<TcpStream> {
     }
 }
 
+async fn connect_via_proxy(proxy_addr: String, addr: NetworkAddress) -> io::Result<TcpStream> {
+    let protos = addr.as_slice();
+
+    if let Some(((host, port), _addr_suffix)) = parse_tcp(protos) {
+        let mut stream = TcpStream::connect(proxy_addr).await?;
+        let mut buffer = [0; 4096];
+        let mut read = 0;
+
+        stream
+            .write_all(&format!("CONNECT {0}:{1} HTTP/1.0\r\n\r\n", host, port).into_bytes())
+            .await?;
+
+        loop {
+            let len = stream.read(&mut buffer[read..]).await?;
+            read += len;
+            let msg = &buffer[..read];
+
+            if len == 0 {
+                break Err(io::Error::new(
+                    io::ErrorKind::Other,
+                    format!(
+                        "HTTP proxy CONNECT failed: {}",
+                        String::from_utf8_lossy(&msg)
+                    ),
+                ));
+            } else if msg.len() >= 16
+                && (msg.starts_with(b"HTTP/1.1 200") || msg.starts_with(b"HTTP/1.0 200"))
+                && msg.ends_with(b"\r\n\r\n")
+            {
+                break Ok(stream);
```

Comment:
I found these `break EXPR;` a bit surprising, didn't realise rust supported this : o

let's just replace them with `return` to keep it simple : )